Analysis/ALICE3/src/alice3-qa-singleparticle.cxx

```diff
@@ -0,0 +1,130 @@
+    histos.add("prodVxVsPt", "Particle Prod. Vertex X " + tit + ";#it{p}_{T} (GeV/#it{c});Prod. Vertex X (cm)", kTH2D, {axisPt, axisProd});
+    histos.add("prodVyVsPt", "Particle Prod. Vertex Y " + tit + ";#it{p}_{T} (GeV/#it{c});Prod. Vertex Y (cm)", kTH2D, {axisPt, axisProd});
+    histos.add("prodVzVsPt", "Particle Prod. Vertex Z " + tit + ";#it{p}_{T} (GeV/#it{c});Prod. Vertex Z (cm)", kTH2D, {axisPt, axisProd});
+    histos.add("prodRadiusVsPt", "Particle Prod. Vertex Radius " + tit + ";#it{p}_{T} (GeV/#it{c});Prod. Vertex Radius (cm)", kTH2D, {axisPt, axisProd});
+    histos.add("prodRadius3DVsPt", "Particle Prod. Vertex Radius XYZ " + tit + ";#it{p}_{T} (GeV/#it{c});Prod. Vertex Radius XYZ (cm)", kTH2D, {axisPt, axisProd});
+    histos.add("trackPt", "Track Pt " + tit + ";#it{p}_{T} (GeV/#it{c})", kTH1D, {axisPt});
+    histos.add("particleEta", "Particle Eta " + tit + ";#it{#eta}", kTH1D, {axisEta});
+    histos.add("trackEta", "Track Eta " + tit + ";#it{#eta}", kTH1D, {axisEta});
+    histos.add("particleY", "Particle Y " + tit + ";#it{y}", kTH1D, {axisEta});
+    histos.add("primaries", "Source for primaries " + tit + ";PDG Code", kTH1D, {{100, 0.f, 100.f}});
+    histos.add("secondaries", "Source for secondaries " + tit + ";PDG Code", kTH1D, {{100, 0.f, 100.f}});
+  }
+
+  void process(const soa::Join<o2::aod::Tracks, o2::aod::McTrackLabels>& tracks,
+               const aod::McParticles& mcParticles)
+  {
+
+    std::vector<int64_t> ParticlesOfInterest;
+    for (const auto& mcParticle : mcParticles) {
+      if (mcParticle.pdgCode() != PDG) {
+        continue;
+      }
+      if (mcParticle.y() < yMin || mcParticle.y() > yMax) {
+        continue;
+      }
+      histos.fill(HIST("particlePt"), mcParticle.pt());
+      histos.fill(HIST("particleEta"), mcParticle.eta());
+      histos.fill(HIST("particleY"), mcParticle.y());
+      histos.fill(HIST("prodVx"), mcParticle.vx());
+      histos.fill(HIST("prodVy"), mcParticle.vy());
+      histos.fill(HIST("prodVz"), mcParticle.vz());
+      histos.fill(HIST("prodRadius"), std::sqrt(mcParticle.vx() * mcParticle.vx() + mcParticle.vy() * mcParticle.vy()));
+      histos.fill(HIST("prodVxVsPt"), mcParticle.pt(), mcParticle.vx());
+      histos.fill(HIST("prodVyVsPt"), mcParticle.pt(), mcParticle.vy());
+      histos.fill(HIST("prodVzVsPt"), mcParticle.pt(), mcParticle.vz());
+      histos.fill(HIST("prodRadiusVsPt"), mcParticle.pt(), std::sqrt(mcParticle.vx() * mcParticle.vx() + mcParticle.vy() * mcParticle.vy()));
+      histos.fill(HIST("prodRadius3DVsPt"), mcParticle.pt(), std::sqrt(mcParticle.vx() * mcParticle.vx() + mcParticle.vy() * mcParticle.vy() + mcParticle.vz() * mcParticle.vz()));
+      ParticlesOfInterest.push_back(mcParticle.globalIndex());
+    }
+
+    for (const auto& track : tracks) {
+      const auto mcParticle = track.mcParticle();
+      if (!IsStable) {
```

Comment:
Just as a side comment, since I have seen it in many places, by many people. I think it's way more readable to have the precondintions for skipping an entry in a loop as early as possible, so that you reduce the amount of nesting you need. in this case, for example, I would have something like:

```
if (!isStable && mcParticle.mother0() < 0) {
  continue;
}
if (isStable && mcParticle.pdgCode() != 0) {
  continue;
}
if (!isStable) {
        auto mother = mcParticles.iteratorAt(mcParticle.mother0());
         const auto ParticleIsInteresting = std::find(ParticlesOfInterest.begin(), ParticlesOfInterest.end(), mother.globalIndex()) != ParticlesOfInterest.end();
         if (ParticleIsInteresting && doPrint) {
           Printf("Track %li comes from a %i and is a %i", track.globalIndex(), mother.pdgCode(), mcParticle.pdgCode());
         }
  continue;
}

... etc...
```

If you want: the more the code is not nested, the easier to read it is.

Reply:
Ciao Giulio, I see, thanks for the suggestion I will try to avoid this in the future